README.md
# tweening-counter

A counter that uses a timing function to move towards it's destination.

[![browser support](https://ci.testling.com/tanem/tweening-counter.png)](https://ci.testling.com/tanem/tweening-counter)

## Live Demo

[Check out the live demo on CodePen](http://codepen.io/tanem/pen/AaxDd), which uses the stand-alone file.

## Installation

Ensure [component(1)](http://component.io) is installed, then:

```
$ component install tanem/tweening-counter
```

## Stand-alone

This library may be used stand-alone without the component tool. To build the stand-alone files, ensure [UglifyJS2](https://github.com/mishoo/UglifyJS2) is installed, then: 

```
$ make standalone
```

Then add ./tweening-counter.js to your application and reference the `TweeningCounter` global.

## Unit Tests

To run the unit tests, ensure [testling](https://github.com/substack/testling) is installed. Then:

```
$ make test
```

## API

### new TweeningCounter(el:Element)

Initialize a new `TweeningCounter` with the given `el`.

### TweeningCounter#to(val:Number)

Set the value to tween to.

```js
tweeningCounter.to(50)
```

### TweeningCounter#ease(fn:String|Function)

Set the easing function.

```js
tweeningCounter.ease('in-out-sine')
tweeningCounter.ease(function(){})
```

### TweeningCounter#duration(ms:Number)

Set the easing duration.

```js
tweeningCounter.duration(1000)
```

### TweeningCounter#onEnd(fn:Function)

Add an `end` event handler bound to this `TweeningCounter`.

```js
tweeningCounter.onEnd(function(){})
```

### TweeningCounter#fixed(fixed:Number)

Specify the number of decimal places in the output.
 
```js
tweeningCounter.fixed(2)
```

### TweeningCounter#start

Start the counter.

```js
tweeningCounter.start()
```

## License

MIT

CHANGELOG.md
## v4.1.4 / 2016-02-07

 * ([7a34273](https://github.com/tanem/tweening-counter/commit/7a3427302229dfb75deb1f5e323736732c3a8703)) Update license (closes [#1](https://github.com/tanem/tweening-counter/issues/1))

## v4.1.3 / 2015-08-27

 * ([36be951](https://github.com/tanem/tweening-counter/commit/36be951024f14b8b8f90fbff2c3ce73590509bac)) Update README
 * ([1d2e90b](https://github.com/tanem/tweening-counter/commit/1d2e90bda3f6c4fa934453667d8c666b5bb7dd8f)) Update license info

## v4.1.2 / 2014-10-05

 * ([732eea1](https://github.com/tanem/tweening-counter/commit/732eea15d8be878052cf47448d658ceb9711b899)) Generate CHANGELOG with mkclog

## v4.1.1 / 2014-07-07

 * ([8bb76ab](https://github.com/tanem/tweening-counter/commit/8bb76ab238fe1833cc994eb473b55cd41efb195f)) Add license

## v4.1.0 / 2014-04-30

 * ([e7b156f](https://github.com/tanem/tweening-counter/commit/e7b156f5bde2dd6e576572bf6c469fc668e25bc1)) Add ability to set number of decimal places in the output

## v4.0.9 / 2014-04-24

 * ([5bec133](https://github.com/tanem/tweening-counter/commit/5bec133e593b99cbc87524024a99aefbe31181a0)) Fix dependency loading

## v4.0.8 / 2014-04-24

 * ([6673587](https://github.com/tanem/tweening-counter/commit/66735875384d3a5ae1d3549462fed6ca6ee59b8f)) Fix component.json syntax

## v4.0.7 / 2014-03-05

 * ([76cc698](https://github.com/tanem/tweening-counter/commit/76cc6985695ac3fee17930da54e279f4e56605dd)) Simplified things around the html template

## v4.0.6 / 2014-03-05

 * ([f81c101](https://github.com/tanem/tweening-counter/commit/f81c101d2bbb1bdb5dc369ea7128e0c7f03f346a)) Update license information

## v4.0.5 / 2014-03-05

 * ([041570a](https://github.com/tanem/tweening-counter/commit/041570acac304f24b6e55fa422fa3b7ab937f556)) Ocd

## v4.0.4 / 2014-03-01

 * ([80c2821](https://github.com/tanem/tweening-counter/commit/80c2821119173c2e791135462f8b8e96f08bc7a8)) Simplify standalone build

## v4.0.3 / 2014-03-01

 * ([f9cf96d](https://github.com/tanem/tweening-counter/commit/f9cf96da33af178c3ecd98d6e636fe8a8ee0f389)) Remove unneeded test file
 * ([1bd4dc9](https://github.com/tanem/tweening-counter/commit/1bd4dc98f39b603c38248b1987d20b9ada43f7c1)) Fix test script
 * ([afabc48](https://github.com/tanem/tweening-counter/commit/afabc48e2a6b047409e6e5013172b45494868a2d)) Add repo field package.json
 * ([bcac0a9](https://github.com/tanem/tweening-counter/commit/bcac0a9cc06bce68388a307ef01228f487f22804)) Update README
 * ([f85050b](https://github.com/tanem/tweening-counter/commit/f85050b47028d9b68af62b38d2c5fea3e5f8679a)) Ocd
 * ([ef778a0](https://github.com/tanem/tweening-counter/commit/ef778a0299b0da74de0af9143fcc917157da4146)) Update test script
 * ([e0bfd92](https://github.com/tanem/tweening-counter/commit/e0bfd925180fb20bbae5425f4c75b93a99b5f568)) Tidy up package.json
 * ([3c3ad78](https://github.com/tanem/tweening-counter/commit/3c3ad78884c85ddcb72c6530560f5587d5dfb3da)) Add mobile browsers for testling
 * ([badd500](https://github.com/tanem/tweening-counter/commit/badd5004a220fa986d1293f7d937ce919ccaae38)) Update README with testling badge and temp addition of build.js for testling
 * ([8c759f3](https://github.com/tanem/tweening-counter/commit/8c759f3254324b6aa0ff6c48d24e49030b1e2cdc)) Ocd

test/helper.js
// Shim for the bind function.
// @see: https://github.com/ariya/phantomjs/issues/10522
// @see: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind
if (!Function.prototype.bind) {
  Function.prototype.bind = function(oThis){
    'use strict';
    if (typeof this !== 'function') {
      // closest thing possible to the ECMAScript 5 internal IsCallable function
      throw new TypeError('Function.prototype.bind - what is trying to be bound is not callable');
    }
 
    var aArgs = Array.prototype.slice.call(arguments, 1),
      fToBind = this,
      FNOP = function(){},
      fBound = function(){
        return fToBind.apply(this instanceof FNOP &&
          oThis ? this : oThis,
          aArgs.concat(Array.prototype.slice.call(arguments)));
      };

    FNOP.prototype = this.prototype;
    fBound.prototype = new FNOP();
 
    return fBound;
  };
}

test/counter.js
describe('Counter', function(){

  'use strict';

  var counter;

  beforeEach(function(){

    var div = document.createElement('div');
    div.className = 'counter';

    counter = new window.Counter({
      el: div
    });

  });

  it('should have a default start value of 0', function(){
    expect(counter.startVal).toBe(0);
  });

  it('should have a default finish value of 0', function(){
    expect(counter.finishVal).toBe(0);
  });

  it('should have a default duration of 3s', function(){
    expect(counter.durationVal).toBe(3);
  });

  it('should throw an error if el is not defined', function(){
    expect(function(){
      counter = new window.Counter();
    }).toThrow('el must be defined');
  });

  it('should set the start value correctly', function(){
    counter.start(10);
    expect(counter.startVal).toBe(10);
  });

  it('should set the finish value correctly', function(){
    counter.finish(10);
    expect(counter.finishVal).toBe(10);
  });

  it('should set the duration value correctly', function(){
    counter.duration(5);
    expect(counter.durationVal).toBe(5);
  });

  it('should default to the correct duration value if an explicit value is not passed', function(){
    counter.duration();
    expect(counter.durationVal).toBe(3);
  });

  describe('Run:', function(){

    beforeEach(function(){
      spyOn(window, 'requestAnimFrame');
      spyOn(Date, 'now').andReturn(1000);
    });

    it('should set the correct delta value', function(){
      counter
        .start(5)
        .finish(10)
        .run();
      expect(counter.delta).toBe(5);
    });

    it('should set the beginning value to the current time', function(){
      counter.run();
      expect(counter.beginning).toBe(1000);
    });

    it('should set the finish callback correctly', function(){
      var cb = function(){};
      counter.run(cb);
      expect(counter.cb).toBe(cb);
    });

    it('should start the animation via requestAnimFrame', function(){
      counter.run();
      expect(window.requestAnimFrame).toHaveBeenCalledWith(counter._animate);
    });

  });

  describe('Animate:', function(){

    beforeEach(function(){
      spyOn(window, 'requestAnimFrame');
      spyOn(counter, 'timingFunction').andCallThrough();
      counter.beginning = 0;
      counter.delta = 50;
      counter.finishVal = 50;
    });

    it('Should call the timing function with the correct arguments', function(){
      spyOn(Date, 'now').andReturn(1000);
      counter._animate();
      expect(counter.timingFunction).toHaveBeenCalledWith(1, 0, 50, 3);
    });

    it('Should set el to the correct value during the animation', function(){
      spyOn(Date, 'now').andReturn(1000);
      counter._animate();
      expect(counter.el.textContent).toBe('35');
    });

    it('Should correctly determine if the animation has finished when moving in the positive direction', function(){
      expect(counter._hasReachedFinish(50)).toBe(true);
    });

    it('Should correctly determine if the animation has finished when moving in the negative direction', function(){
      counter.delta = -50;
      counter.finishVal = 0;

      expect(counter._hasReachedFinish(0)).toBe(true);
    });

    it('Should correctly determine if the animation has finished when there is no difference between start and finish values', function(){
      counter.delta = 0;
      expect(counter._hasReachedFinish(0)).toBe(true);
    });

    it('Should call any provided callback when the animation has finished', function(){
      spyOn(Date, 'now').andReturn(3000);
      counter.cb = jasmine.createSpy();

      counter._animate();

      expect(counter.cb).toHaveBeenCalled();
    });

    it('Should set el to the finish value when the animation has finished', function(){
      spyOn(Date, 'now').andReturn(3000);
      counter._animate();
      expect(counter.el.textContent).toBe('50');
    });

  });

});


standalone/tweening-counter.js
;(function(){

/**
 * Require the given path.
 *
 * @param {String} path
 * @return {Object} exports
 * @api public
 */

function require(path, parent, orig) {
  var resolved = require.resolve(path);

  // lookup failed
  if (null == resolved) {
    orig = orig || path;
    parent = parent || 'root';
    var err = new Error('Failed to require "' + orig + '" from "' + parent + '"');
    err.path = orig;
    err.parent = parent;
    err.require = true;
    throw err;
  }

  var module = require.modules[resolved];

  // perform real require()
  // by invoking the module's
  // registered function
  if (!module._resolving && !module.exports) {
    var mod = {};
    mod.exports = {};
    mod.client = mod.component = true;
    module._resolving = true;
    module.call(this, mod.exports, require.relative(resolved), mod);
    delete module._resolving;
    module.exports = mod.exports;
  }

  return module.exports;
}

/**
 * Registered modules.
 */

require.modules = {};

/**
 * Registered aliases.
 */

require.aliases = {};

/**
 * Resolve `path`.
 *
 * Lookup:
 *
 *   - PATH/index.js
 *   - PATH.js
 *   - PATH
 *
 * @param {String} path
 * @return {String} path or null
 * @api private
 */

require.resolve = function(path) {
  if (path.charAt(0) === '/') path = path.slice(1);

  var paths = [
    path,
    path + '.js',
    path + '.json',
    path + '/index.js',
    path + '/index.json'
  ];

  for (var i = 0; i < paths.length; i++) {
    var path = paths[i];
    if (require.modules.hasOwnProperty(path)) return path;
    if (require.aliases.hasOwnProperty(path)) return require.aliases[path];
  }
};

/**
 * Normalize `path` relative to the current path.
 *
 * @param {String} curr
 * @param {String} path
 * @return {String}
 * @api private
 */

require.normalize = function(curr, path) {
  var segs = [];

  if ('.' != path.charAt(0)) return path;

  curr = curr.split('/');
  path = path.split('/');

  for (var i = 0; i < path.length; ++i) {
    if ('..' == path[i]) {
      curr.pop();
    } else if ('.' != path[i] && '' != path[i]) {
      segs.push(path[i]);
    }
  }

  return curr.concat(segs).join('/');
};

/**
 * Register module at `path` with callback `definition`.
 *
 * @param {String} path
 * @param {Function} definition
 * @api private
 */

require.register = function(path, definition) {
  require.modules[path] = definition;
};

/**
 * Alias a module definition.
 *
 * @param {String} from
 * @param {String} to
 * @api private
 */

require.alias = function(from, to) {
  if (!require.modules.hasOwnProperty(from)) {
    throw new Error('Failed to alias "' + from + '", it does not exist');
  }
  require.aliases[to] = from;
};

/**
 * Return a require function relative to the `parent` path.
 *
 * @param {String} parent
 * @return {Function}
 * @api private
 */

require.relative = function(parent) {
  var p = require.normalize(parent, '..');

  /**
   * lastIndexOf helper.
   */

  function lastIndexOf(arr, obj) {
    var i = arr.length;
    while (i--) {
      if (arr[i] === obj) return i;
    }
    return -1;
  }

  /**
   * The relative require() itself.
   */

  function localRequire(path) {
    var resolved = localRequire.resolve(path);
    return require(resolved, parent, path);
  }

  /**
   * Resolve relative to the parent.
   */

  localRequire.resolve = function(path) {
    var c = path.charAt(0);
    if ('/' == c) return path.slice(1);
    if ('.' == c) return require.normalize(p, path);

    // resolve deps by returning
    // the dep in the nearest "deps"
    // directory
    var segs = parent.split('/');
    var i = lastIndexOf(segs, 'deps') + 1;
    if (!i) i = 0;
    path = segs.slice(0, i + 1).join('/') + '/deps/' + path;
    return path;
  };

  /**
   * Check if module is defined at `path`.
   */

  localRequire.exists = function(path) {
    return require.modules.hasOwnProperty(localRequire.resolve(path));
  };

  return localRequire;
};
require.register("component-raf/index.js", function(exports, require, module){
/**
 * Expose `requestAnimationFrame()`.
 */

exports = module.exports = window.requestAnimationFrame
  || window.webkitRequestAnimationFrame
  || window.mozRequestAnimationFrame
  || window.oRequestAnimationFrame
  || window.msRequestAnimationFrame
  || fallback;

/**
 * Fallback implementation.
 */

var prev = new Date().getTime();
function fallback(fn) {
  var curr = new Date().getTime();
  var ms = Math.max(0, 16 - (curr - prev));
  var req = setTimeout(fn, ms);
  prev = curr;
  return req;
}

/**
 * Cancel.
 */

var cancel = window.cancelAnimationFrame
  || window.webkitCancelAnimationFrame
  || window.mozCancelAnimationFrame
  || window.oCancelAnimationFrame
  || window.msCancelAnimationFrame
  || window.clearTimeout;

exports.cancel = function(id){
  cancel.call(window, id);
};

});
require.register("segmentio-extend/index.js", function(exports, require, module){

module.exports = function extend (object) {
    // Takes an unlimited number of extenders.
    var args = Array.prototype.slice.call(arguments, 1);

    // For each extender, copy their properties on our object.
    for (var i = 0, source; source = args[i]; i++) {
        if (!source) continue;
        for (var property in source) {
            object[property] = source[property];
        }
    }

    return object;
};
});
require.register("techjacker-isfunction/lib/main.js", function(exports, require, module){
// if (typeof require !== 'undefined') {}

var isFunction = function (functionToCheck) {
	var getType = {};
	return functionToCheck && getType.toString.call(functionToCheck) === '[object Function]';
};

if (typeof module !== 'undefined' && module.exports) {
	module.exports = isFunction;
}
});
require.register("component-bind/index.js", function(exports, require, module){
/**
 * Slice reference.
 */

var slice = [].slice;

/**
 * Bind `obj` to `fn`.
 *
 * @param {Object} obj
 * @param {Function|String} fn or string
 * @return {Function}
 * @api public
 */

module.exports = function(obj, fn){
  if ('string' == typeof fn) fn = obj[fn];
  if ('function' != typeof fn) throw new Error('bind() requires a function');
  var args = slice.call(arguments, 2);
  return function(){
    return fn.apply(obj, args.concat(slice.call(arguments)));
  }
};

});
require.register("component-emitter/index.js", function(exports, require, module){

/**
 * Expose `Emitter`.
 */

module.exports = Emitter;

/**
 * Initialize a new `Emitter`.
 *
 * @api public
 */

function Emitter(obj) {
  if (obj) return mixin(obj);
};

/**
 * Mixin the emitter properties.
 *
 * @param {Object} obj
 * @return {Object}
 * @api private
 */

function mixin(obj) {
  for (var key in Emitter.prototype) {
    obj[key] = Emitter.prototype[key];
  }
  return obj;
}

/**
 * Listen on the given `event` with `fn`.
 *
 * @param {String} event
 * @param {Function} fn
 * @return {Emitter}
 * @api public
 */

Emitter.prototype.on = function(event, fn){
  this._callbacks = this._callbacks || {};
  (this._callbacks[event] = this._callbacks[event] || [])
    .push(fn);
  return this;
};

/**
 * Adds an `event` listener that will be invoked a single
 * time then automatically removed.
 *
 * @param {String} event
 * @param {Function} fn
 * @return {Emitter}
 * @api public
 */

Emitter.prototype.once = function(event, fn){
  var self = this;
  this._callbacks = this._callbacks || {};

  function on() {
    self.off(event, on);
    fn.apply(this, arguments);
  }

  fn._off = on;
  this.on(event, on);
  return this;
};

/**
 * Remove the given callback for `event` or all
 * registered callbacks.
 *
 * @param {String} event
 * @param {Function} fn
 * @return {Emitter}
 * @api public
 */

Emitter.prototype.off =
Emitter.prototype.removeListener =
Emitter.prototype.removeAllListeners = function(event, fn){
  this._callbacks = this._callbacks || {};

  // all
  if (0 == arguments.length) {
    this._callbacks = {};
    return this;
  }

  // specific event
  var callbacks = this._callbacks[event];
  if (!callbacks) return this;

  // remove all handlers
  if (1 == arguments.length) {
    delete this._callbacks[event];
    return this;
  }

  // remove specific handler
  var i = callbacks.indexOf(fn._off || fn);
  if (~i) callbacks.splice(i, 1);
  return this;
};

/**
 * Emit `event` with the given args.
 *
 * @param {String} event
 * @param {Mixed} ...
 * @return {Emitter}
 */

Emitter.prototype.emit = function(event){
  this._callbacks = this._callbacks || {};
  var args = [].slice.call(arguments, 1)
    , callbacks = this._callbacks[event];

  if (callbacks) {
    callbacks = callbacks.slice(0);
    for (var i = 0, len = callbacks.length; i < len; ++i) {
      callbacks[i].apply(this, args);
    }
  }

  return this;
};

/**
 * Return array of callbacks for `event`.
 *
 * @param {String} event
 * @return {Array}
 * @api public
 */

Emitter.prototype.listeners = function(event){
  this._callbacks = this._callbacks || {};
  return this._callbacks[event] || [];
};

/**
 * Check if this emitter has `event` handlers.
 *
 * @param {String} event
 * @return {Boolean}
 * @api public
 */

Emitter.prototype.hasListeners = function(event){
  return !! this.listeners(event).length;
};

});
require.register("component-tween/index.js", function(exports, require, module){

/**
 * Module dependencies.
 */

var Emitter = require('emitter')
  , ease = require('ease');

/**
 * Expose `Tween`.
 */

module.exports = Tween;

/**
 * Initialize a new `Tween` with `obj`.
 *
 * @param {Object|Array} obj
 * @api public
 */

function Tween(obj) {
  if (!(this instanceof Tween)) return new Tween(obj);
  this._from = obj;
  this.ease('linear');
  this.duration(500);
}

/**
 * Mixin emitter.
 */

Emitter(Tween.prototype);

/**
 * Reset the tween.
 *
 * @api public
 */

Tween.prototype.reset = function(){
  this.isArray = Array.isArray(this._from);
  this._curr = clone(this._from);
  this._done = false;
  this._start = Date.now();
  return this;
};

/**
 * Tween to `obj` and reset internal state.
 *
 *    tween.to({ x: 50, y: 100 })
 *
 * @param {Object|Array} obj
 * @return {Tween} self
 * @api public
 */

Tween.prototype.to = function(obj){
  this.reset();
  this._to = obj;
  return this;
};

/**
 * Set duration to `ms` [500].
 *
 * @param {Number} ms
 * @return {Tween} self
 * @api public
 */

Tween.prototype.duration = function(ms){
  this._duration = ms;
  return this;
};

/**
 * Set easing function to `fn`.
 *
 *    tween.ease('in-out-sine')
 *
 * @param {String|Function} fn
 * @return {Tween}
 * @api public
 */

Tween.prototype.ease = function(fn){
  fn = 'function' == typeof fn ? fn : ease[fn];
  if (!fn) throw new TypeError('invalid easing function');
  this._ease = fn;
  return this;
};

/**
 * Stop the tween and immediately emit "stop" and "end".
 *
 * @return {Tween}
 * @api public
 */

Tween.prototype.stop = function(){
  this.stopped = true;
  this._done = true;
  this.emit('stop');
  this.emit('end');
  return this;
};

/**
 * Perform a step.
 *
 * @return {Tween} self
 * @api private
 */

Tween.prototype.step = function(){
  if (this._done) return;

  // duration
  var duration = this._duration;
  var now = Date.now();
  var delta = now - this._start;
  var done = delta >= duration;

  // complete
  if (done) {
    this._from = this._to;
    this._update(this._to);
    this._done = true;
    this.emit('end');
    return this;
  }

  // tween
  var from = this._from;
  var to = this._to;
  var curr = this._curr;
  var fn = this._ease;
  var p = (now - this._start) / duration;
  var n = fn(p);

  // array
  if (this.isArray) {
    for (var i = 0; i < from.length; ++i) {
      curr[i] = from[i] + (to[i] - from[i]) * n;
    }

    this._update(curr);
    return this;
  }

  // objech
  for (var k in from) {
    curr[k] = from[k] + (to[k] - from[k]) * n;
  }

  this._update(curr);
  return this;
};

/**
 * Set update function to `fn` or
 * when no argument is given this performs
 * a "step".
 *
 * @param {Function} fn
 * @return {Tween} self
 * @api public
 */

Tween.prototype.update = function(fn){
  if (0 == arguments.length) return this.step();
  this._update = fn;
  return this;
};

/**
 * Clone `obj`.
 *
 * @api private
 */

function clone(obj) {
  if (Array.isArray(obj)) return obj.slice();
  var ret = {};
  for (var key in obj) ret[key] = obj[key];
  return ret;
}

});
require.register("component-ease/index.js", function(exports, require, module){

exports.linear = function(n){
  return n;
};

exports.inQuad = function(n){
  return n * n;
};

exports.outQuad = function(n){
  return n * (2 - n);
};

exports.inOutQuad = function(n){
  n *= 2;
  if (n < 1) return 0.5 * n * n;
  return - 0.5 * (--n * (n - 2) - 1);
};

exports.inCube = function(n){
  return n * n * n;
};

exports.outCube = function(n){
  return --n * n * n + 1;
};

exports.inOutCube = function(n){
  n *= 2;
  if (n < 1) return 0.5 * n * n * n;
  return 0.5 * ((n -= 2 ) * n * n + 2);
};

exports.inQuart = function(n){
  return n * n * n * n;
};

exports.outQuart = function(n){
  return 1 - (--n * n * n * n);
};

exports.inOutQuart = function(n){
  n *= 2;
  if (n < 1) return 0.5 * n * n * n * n;
  return -0.5 * ((n -= 2) * n * n * n - 2);
};

exports.inQuint = function(n){
  return n * n * n * n * n;
}

exports.outQuint = function(n){
  return --n * n * n * n * n + 1;
}

exports.inOutQuint = function(n){
  n *= 2;
  if (n < 1) return 0.5 * n * n * n * n * n;
  return 0.5 * ((n -= 2) * n * n * n * n + 2);
};

exports.inSine = function(n){
  return 1 - Math.cos(n * Math.PI / 2 );
};

exports.outSine = function(n){
  return Math.sin(n * Math.PI / 2);
};

exports.inOutSine = function(n){
  return .5 * (1 - Math.cos(Math.PI * n));
};

exports.inExpo = function(n){
  return 0 == n ? 0 : Math.pow(1024, n - 1);
};

exports.outExpo = function(n){
  return 1 == n ? n : 1 - Math.pow(2, -10 * n);
};

exports.inOutExpo = function(n){
  if (0 == n) return 0;
  if (1 == n) return 1;
  if ((n *= 2) < 1) return .5 * Math.pow(1024, n - 1);
  return .5 * (-Math.pow(2, -10 * (n - 1)) + 2);
};

exports.inCirc = function(n){
  return 1 - Math.sqrt(1 - n * n);
};

exports.outCirc = function(n){
  return Math.sqrt(1 - (--n * n));
};

exports.inOutCirc = function(n){
  n *= 2
  if (n < 1) return -0.5 * (Math.sqrt(1 - n * n) - 1);
  return 0.5 * (Math.sqrt(1 - (n -= 2) * n) + 1);
};

exports.inBack = function(n){
  var s = 1.70158;
  return n * n * (( s + 1 ) * n - s);
};

exports.outBack = function(n){
  var s = 1.70158;
  return --n * n * ((s + 1) * n + s) + 1;
};

exports.inOutBack = function(n){
  var s = 1.70158 * 1.525;
  if ( ( n *= 2 ) < 1 ) return 0.5 * ( n * n * ( ( s + 1 ) * n - s ) );
  return 0.5 * ( ( n -= 2 ) * n * ( ( s + 1 ) * n + s ) + 2 );
};

exports.inBounce = function(n){
  return 1 - exports.outBounce(1 - n);
};

exports.outBounce = function(n){
  if ( n < ( 1 / 2.75 ) ) {
    return 7.5625 * n * n;
  } else if ( n < ( 2 / 2.75 ) ) {
    return 7.5625 * ( n -= ( 1.5 / 2.75 ) ) * n + 0.75;
  } else if ( n < ( 2.5 / 2.75 ) ) {
    return 7.5625 * ( n -= ( 2.25 / 2.75 ) ) * n + 0.9375;
  } else {
    return 7.5625 * ( n -= ( 2.625 / 2.75 ) ) * n + 0.984375;
  }
};

exports.inOutBounce = function(n){
  if (n < .5) return exports.inBounce(n * 2) * .5;
  return exports.outBounce(n * 2 - 1) * .5 + .5;
};

// aliases

exports['in-quad'] = exports.inQuad;
exports['out-quad'] = exports.outQuad;
exports['in-out-quad'] = exports.inOutQuad;
exports['in-cube'] = exports.inCube;
exports['out-cube'] = exports.outCube;
exports['in-out-cube'] = exports.inOutCube;
exports['in-quart'] = exports.inQuart;
exports['out-quart'] = exports.outQuart;
exports['in-out-quart'] = exports.inOutQuart;
exports['in-quint'] = exports.inQuint;
exports['out-quint'] = exports.outQuint;
exports['in-out-quint'] = exports.inOutQuint;
exports['in-sine'] = exports.inSine;
exports['out-sine'] = exports.outSine;
exports['in-out-sine'] = exports.inOutSine;
exports['in-expo'] = exports.inExpo;
exports['out-expo'] = exports.outExpo;
exports['in-out-expo'] = exports.inOutExpo;
exports['in-circ'] = exports.inCirc;
exports['out-circ'] = exports.outCirc;
exports['in-out-circ'] = exports.inOutCirc;
exports['in-back'] = exports.inBack;
exports['out-back'] = exports.outBack;
exports['in-out-back'] = exports.inOutBack;
exports['in-bounce'] = exports.inBounce;
exports['out-bounce'] = exports.outBounce;
exports['in-out-bounce'] = exports.inOutBounce;

});
require.register("component-domify/index.js", function(exports, require, module){

/**
 * Expose `parse`.
 */

module.exports = parse;

/**
 * Wrap map from jquery.
 */

var map = {
  legend: [1, '<fieldset>', '</fieldset>'],
  tr: [2, '<table><tbody>', '</tbody></table>'],
  col: [2, '<table><tbody></tbody><colgroup>', '</colgroup></table>'],
  _default: [0, '', '']
};

map.td =
map.th = [3, '<table><tbody><tr>', '</tr></tbody></table>'];

map.option =
map.optgroup = [1, '<select multiple="multiple">', '</select>'];

map.thead =
map.tbody =
map.colgroup =
map.caption =
map.tfoot = [1, '<table>', '</table>'];

map.text =
map.circle =
map.ellipse =
map.line =
map.path =
map.polygon =
map.polyline =
map.rect = [1, '<svg xmlns="http://www.w3.org/2000/svg" version="1.1">','</svg>'];

/**
 * Parse `html` and return the children.
 *
 * @param {String} html
 * @return {Array}
 * @api private
 */

function parse(html) {
  if ('string' != typeof html) throw new TypeError('String expected');
  
  // tag name
  var m = /<([\w:]+)/.exec(html);
  if (!m) return document.createTextNode(html);

  html = html.replace(/^\s+|\s+$/g, ''); // Remove leading/trailing whitespace

  var tag = m[1];

  // body support
  if (tag == 'body') {
    var el = document.createElement('html');
    el.innerHTML = html;
    return el.removeChild(el.lastChild);
  }

  // wrap map
  var wrap = map[tag] || map._default;
  var depth = wrap[0];
  var prefix = wrap[1];
  var suffix = wrap[2];
  var el = document.createElement('div');
  el.innerHTML = prefix + html + suffix;
  while (depth--) el = el.lastChild;

  // one element
  if (el.firstChild == el.lastChild) {
    return el.removeChild(el.firstChild);
  }

  // several elements
  var fragment = document.createDocumentFragment();
  while (el.firstChild) {
    fragment.appendChild(el.removeChild(el.firstChild));
  }

  return fragment;
}

});
require.register("tweening-counter/index.js", function(exports, require, module){
'use strict';

var raf = require('raf');
var Tween = require('tween');
var bind = require('bind');
var isfunction = require('isfunction');
var template = require('./template.html');
var domify = require('domify');

/**
 * Expose `TweeningCounter`.
 */

module.exports = TweeningCounter;

/**
 * Initialize a new `TweeningCounter`.
 *
 * @api public
 */

function TweeningCounter(){
  this.el = domify(template);
  this.interpolate = bind(this, 'interpolate');
  this.update = bind(this, 'update');
  this.cancelAnimationFrame = bind(this, 'cancelAnimationFrame');
  this.tween = Tween({ val: 0 })
    .ease('out-cube')
    .to({ val: 0 })
    .update(this.update)
    .duration(1000)
    .on('end', this.cancelAnimationFrame);
}

/**
 * Set the value to tween to.
 *
 * ```js
 * tweeningCounter.to(50)
 * ```
 *
 * @param {Number} val
 * @return {TweeningCounter} self
 * @api public
 */

TweeningCounter.prototype.to = function(val){
  this.tween.to({ val: val });
  return this;
};

/**
 * Set the easing function.
 *
 * ```js
 * tweeningCounter.ease('in-out-sine')
 * tweeningCounter.ease(function(){})
 * ```
 *
 * @param {String|Function} fn
 * @return {TweeningCounter} self
 * @see component/ease
 * @api public
 */

TweeningCounter.prototype.ease = function(fn){
  this.tween.ease(fn);
  return this;
};

/**
 * Set the easing duration.
 *
 * ```js
 * tweeningCounter.duration(1000);
 * ```
 *
 * @param {Number} ms
 * @return {TweeningCounter} self
 * @api public
 */

TweeningCounter.prototype.duration = function(ms){
  this.tween.duration(ms);
  return this;
};

/**
 * Add an `end` event handler bound to this `TweeningCounter`.
 *
 * ```js
 * tweeningCounter.onEnd(function(){})
 * ```
 *
 * @param {Function} fn
 * @return {TweeningCounter} self
 * @api public
 */

TweeningCounter.prototype.onEnd = function(fn){
  if (isfunction(fn)) this.tween.on('end', bind(this, fn));
  return this;
};

/**
 * Start the counter.
 *
 * ```js
 * tweeningCounter.start()
 * ```
 *
 * @api public
 */

TweeningCounter.prototype.start = function(){
  this.interpolate();
};

/**
 * Perform steps.
 *
 * @api private
 */

TweeningCounter.prototype.interpolate = function(){
  this.id = raf(this.interpolate);
  this.tween.update();
};

/**
 * Set `el` text content to the correct value.
 *
 * @param {Object} obj Contains the value at this point in time.
 * @api private
 */

TweeningCounter.prototype.update = function(obj){
  this.el.textContent = Math.round(obj.val);
};

/**
 * Cancels the last request animation frame.
 *
 * @api private
 */

TweeningCounter.prototype.cancelAnimationFrame = function(){
  raf.cancel(this.rafId);
};
});














require.register("tweening-counter/template.html", function(exports, require, module){
module.exports = '<div class="tweening-counter"></div>';
});
require.alias("component-raf/index.js", "tweening-counter/deps/raf/index.js");
require.alias("component-raf/index.js", "raf/index.js");

require.alias("segmentio-extend/index.js", "tweening-counter/deps/extend/index.js");
require.alias("segmentio-extend/index.js", "extend/index.js");

require.alias("techjacker-isfunction/lib/main.js", "tweening-counter/deps/isfunction/lib/main.js");
require.alias("techjacker-isfunction/lib/main.js", "tweening-counter/deps/isfunction/index.js");
require.alias("techjacker-isfunction/lib/main.js", "isfunction/index.js");
require.alias("techjacker-isfunction/lib/main.js", "techjacker-isfunction/index.js");
require.alias("component-bind/index.js", "tweening-counter/deps/bind/index.js");
require.alias("component-bind/index.js", "bind/index.js");

require.alias("component-tween/index.js", "tweening-counter/deps/tween/index.js");
require.alias("component-tween/index.js", "tween/index.js");
require.alias("component-emitter/index.js", "component-tween/deps/emitter/index.js");

require.alias("component-ease/index.js", "component-tween/deps/ease/index.js");

require.alias("component-ease/index.js", "tweening-counter/deps/ease/index.js");
require.alias("component-ease/index.js", "ease/index.js");

require.alias("component-domify/index.js", "tweening-counter/deps/domify/index.js");
require.alias("component-domify/index.js", "domify/index.js");

require.alias("tweening-counter/index.js", "tweening-counter/index.js");if (typeof exports == "object") {
  module.exports = require("tweening-counter");
} else if (typeof define == "function" && define.amd) {
  define([], function(){ return require("tweening-counter"); });
} else {
  this["TweeningCounter"] = require("tweening-counter");
}})();